[toc]

# JS基础
## 异步加载
### async和defer的区别
相同点
+ 加载文件时不阻塞页面渲染
+ 使用这两个属性的脚本中不能调用document.write方法
+ 有脚本的onload的事件回调

不同点

```js
JS引擎进行此法分析构建DOM树，到script的时候会停下来去加载JS代码，async就是在DOM解析过程异步加载js,加载完就执行，而defer要等DOM解析完成后，再去下载
```

+ 对于defer，我们可以认为是将外链的js放在了页面底部。js的加载不会阻塞页面的渲染和资源的加载。不过defer会按照原本的js的顺序执行，所以如果前后有依赖关系的js可以放心使用。
+ 对于async，这个是html5中新增的属性，它的作用是能够异步的加载和执行脚本，不因为加载脚本而阻塞页面的加载。一旦加载到就会立刻执行在有async的情况下，js一旦下载好了就会执行，所以很有可能不是按照原本的顺序来执行的。如果js前后有依赖性，用async，就很有可能出错。

******
## JS 异步解决方案的发展历程以及优缺点。
+ 1.回调函数（callback）


```js
ajax('XXX1', () => {
    // callback 函数体
    ajax('XXX2', () => {
        // callback 函数体
        ajax('XXX3', () => {
            // callback 函数体
        })
    })
})
```
优点：解决了同步的问题（只要有一个任务耗时很长，后面的任务都必须排队等着，会拖延整个程序的执行。）

缺点：回调地狱，不能用 try catch 捕获错误，不能 return，回调地狱的根本问题在于：


```js
缺乏顺序性： 回调地狱导致的调试困难，和大脑的思维方式不符
嵌套函数存在耦合性，一旦有所改动，就会牵一发而动全身
嵌套函数过多的多话，很难处理错误
```
+ 2.Promise

Promise就是为了解决callback的问题而产生的。
Promise 实现了链式调用，也就是说每次 then 后返回的都是一个全新 Promise，如果我们在 then 中 return ，return 的结果会被 Promise.resolve() 包装


优点：解决了回调地狱的问题

缺点：无法取消 Promise ，错误需要通过回调函数来捕获

```js
ajax('XXX1')
  .then(res => {
      // 操作逻辑
      return ajax('XXX2')
  }).then(res => {
      // 操作逻辑
      return ajax('XXX3')
  }).then(res => {
      // 操作逻辑
  })
```
+ 3.Generator

特点：可以控制函数的执行，可以配合 co 函数库使用

```js
function *fetch() {
    yield ajax('XXX1', () => {})
    yield ajax('XXX2', () => {})
    yield ajax('XXX3', () => {})
}
let it = fetch()
let result1 = it.next()
let result2 = it.next()
let result3 = it.next()
```
+ 4.Async/await

async、await 是异步的终极解决方案

优点是：代码清晰，不用像 Promise 写一大堆 then 链，处理了回调地狱的问题

缺点：await 将异步代码改造成同步代码，如果**多个异步操作没有依赖性**而使用 await 会导致性能上的降低。


```js
async function test() {
  // 以下代码没有依赖性的话，完全可以使用 Promise.all 的方式
  // 如果有依赖性的话，其实就是解决回调地狱的例子了
  await fetch('XXX1')
  await fetch('XXX2')
  await fetch('XXX3')
}
```
**********
## setTimeout、Promise、Async/Await 的区别
+ settimeout的回调函数放到宏任务队列里，等到执行栈清空以后执行；
+ promise.then里的回调函数会放到相应宏任务的微任务队列里，等宏任务里面的同步代码执行完再执行；
+ async函数表示函数里面可能会有异步方法，await后面跟一个表达式，async方法执行时，遇到await会立即执行表达式，然后把表达式后面的代码放到微任务队列里，让出执行栈让同步代码先执行。
**********


## 异步编程
异步编程的语法目标，就是怎样让它更像同步编程。
主要方式：
+ 回调函数
+ 事件监听
+ 发布/订阅
+ Promise

+ Generator

### 1.并发和并行区别
> 并发：宏观上同时完成任务，实质上是通过任务之间的切换完成的

> 并行：实际上就是同时完成多个任务

### 2.回调函数（Callback）
回调函数就是把任务分为两段，第一段执行结束之后才能执行第二段，中间的等待过程，可以执行其它操作，
存在的问题

+ 回调地狱,需要自己去理清楚执行顺序
+ 回调是否执行，执行了多少次,是同步还是异步执行的
+ 错误难以获取，不能直接使用try...catch(因为异步操作，只有一个执行上下文，难以寻找错误)
+ 不能直接return
***************************
### 3.Promise基本使用
解决回调的部分问题
+ 回调执行多次===> Promise的resolve只能执行一次
+ 回调函数有没有执行===> Promise.race
+ 无法判定回调是否被执行(内部混杂了同步异步的函数，比如如果有缓存，执行同步，没有缓存就去请求，执行异步，混杂向外只暴露一个接口)===>then里规定了异步执行顺序

#### 如何中断promise的执行
##### 第一个then强行报错
一种方法是在then 1中直接抛错, 这样就不会执行then 2, then 3, 直接跳到catch方法打印err(但此方法并没有实际中断)

```js
Promise.resolve().then(() => {
    console.log('then 1')
    throw new Error('xxx')
}).then(() => {
    console.log('then 2')
}).then(() => {
    console.log('then 3')
}).catch((err) => {
    console.log(err)
})
```
##### then 1中return 一个新的Promise
在then 1中return 一个新的Promise,但不改变其状态,这样该Promise就一直处于pedding状态,即不会执行后面任何方法，因为Promises/A+标准：原Promise对象的状态将跟新对象保持一致。


```js
Promise.resolve().then(() => {
    console.log('then 1')
    return new Promise(() => {})
}).then(() => {
    console.log('then 2')
}).then(() => {
    console.log('then 3')
}).catch((err) => {
    console.log(err)
})
```
##### Promise.race竞速方法
让其它不执行

```js
let p1 = new Promise((resolve, reject) => {
    resolve('ok1')
})
 
let p2 = new Promise((resolve, reject) => {
    setTimeout(() => {resolve('ok2')}, 10)
})
 
Promise.race([p2, p1]).then((result) => {
    console.log(result) //ok1
}).catch((error) => {
    console.log(error)
})
```



#### 基本使用

```
new Promise((resolve, reject) => {
  resolve('success')
  // 无效
  reject('reject')
})
```

#### catch
+ 捕获then的前一步里的错误，then的reject和catch作用相同
+ 捕获then里的错误，需要在下一个then或者catch里

```js
// bad
somethingAync.then(function() {
    return somethingElseAsync();
}, function(err) {
    handleMyError(err);
});
```

```js
// good
somethingAsync
.then(function() {
    return somethingElseAsync()
})
.then(null, function(err) {
    handleMyError(err);
});
```

```js
// good
somethingAsync()
.then(function() {
    return somethingElseAsync();
})
.catch(function(err) {
    handleMyError(err);
});
```
#### promise.then
promise的then，如果有return，会被promise.resolve包裹，返回依然为promise
#### promise.all

只有全部都执行才能执行，如下，如果p1,p2,p3都执行resolve，那么结果为[1,2,3],如果把p2改为reject(2),那么执行结果为2

```js
const p1 = new Promise((resolve, reject) => {
    resolve(1);
});

const p2 = new Promise((resolve, reject) => {
    resolve(2);
});

const p3 = new Promise((resolve, reject) => {
    resolve(3);
});

Promise.all([p1, p2, p3]).then(data => { 
    console.log(data); // [1, 2, 3] 结果顺序和promise实例数组顺序是一致的
}, err => {
    console.log(err);
});
```
#### Promise.race
只要有一个为rejected或者resolved就执行后续操作

```js
function timerPromisefy(delay) {
    return new Promise(function (resolve, reject) {
        setTimeout(function () {
            resolve(delay);
        }, delay);
    });
}
var startDate = Date.now();

Promise.race([
    timerPromisefy(10),
    timerPromisefy(20),
    timerPromisefy(30)
]).then(function (values) {
    console.log(values); // 10
});
```
#### Promise的缺点
+ 无法捕获错误

```js
// 想要的结果，promise出错，不能被打印
const promise = new Promise(null);
console.log(233333);
//实际上，promise内部的错误无法影响外部，会打印2333333
let promise = new Promise(() => {
    throw new Error('error')
});
console.log(2333333);
```
+ 单一值

```js
//每次向then传递的都为单一的值，需要封装，解封
Promise.all([Promise.resolve(1), Promise.resolve(2)])
.then(([x, y]) => {
    console.log(x, y);
});
```
+ 无法取消

```
一旦开始，无法取消
```


***********************
### 4.手写Promise
promise翻译是承诺，它有三种状态，状态只能从pending变为resolved或者rejected
+ pending 等待中
+ fulfilled已执行
+ rejected 已拒绝

#### 基本Promise
记忆：
+ const that=this
+ try{fn(resolve,reject)}catch(e){reject(e)}
+ 使用 new Promise((resolve,reject)=>{resolve(100)})
```
const PENDING='pending'
const RESOLVED='resolved'
const REJECTED='rejected'
function Mypromise(fn){
    const that = this
    that.state= PENDING
    that.resolvedcallback=[]
    that.rejectedcallback=[]
    that.value=null

    function resolve(value){
        if(that.state=== PENDING){
            that.state=RESOLVED
            that.value=value
            that.resolvedcallback.map((cb)=>cb(that.value))
        }
        console.log('resolve',that.value)
    }
    function reject(value){
        if(that.state === PENDING){
            that.state=REJECTED
            that.value=value
            that.rejectedcallback.map((cb)=>cb(that.value))
        }
        console.log('reject',that.value)
    }            
    try{
        fn(resolve,reject)
    }catch(e){
        reject(e)
    }
}

var peomise1=new Mypromise((resolve,reject)=>{
    resolve(100)
    reject(200)
})
```
+ then

```
观察者模式
then收集依赖，放入回调函数数组里 ->
异步触发resolve -> resolve执行依赖
```


```
Mypromise.prototype.then=function(onfullfilled,onrejected){
    const that =this
    if(that.state===PENDING){
        that.resolvedcallback.push(onfullfilled)
        that.rejectedcallback.push(onrejected)
    }
    if(that.state===RESOLVED){
        onfullfilled(that.value)
    }
    if(that.state===REJECTED){
        onrejected(that.value)
    }
}

var peomise1=new Mypromise((resolve,reject)=>{
    resolve(100)
    reject(200)
}).then(data=>{console.log(data)})
```

##### promise状态透传：
promise A+规定 then方法必须返回一个promise实例
+ 如果 promise1的 onFulfilled 不是函数，那么promise1的不可变值将传递到promise2并作为promise2的不可变值。
+ 如果 promise1的 onRejected不是函数，那么promise1的不可变原因将传递到promise2并作为promise2的不可变原因，并作为promise2的 onRejected 的入参。
+ 这就是Promise的==状态透传==特点，如果当前的promise实例没有绑定回调函数，或者绑定的不是函数，那么当前实例就会把其状态以及不可变值或者不可变原因传递给当前实例调用.then方法返回的新promise实例。==》使用最后的catch捕获之前没有进行处理的rejected状态。
```
new MyPromise((resolve, reject) => { 
    setTimeout(() => { 
         resolve(1) 
    }, 0) 
}).then(value => { console.log('then',value) })

//结果：then pending
then 1
resolve 1
因为setTimeout异步，所以状态为等待状态，先把then里的函数传递到回调函数数组，先执行了then所传递的函数，才执行promise本身的函数
```
#### Promise A+
+ 改造resolve和reject函数，resolve判断是否为promise类型，是则执行then
+ resolve和reject用setTimeout包裹

```
//resolve函数
    function resolve(value) { 
       if (value instanceof MyPromise) {//首先需要判断传入的值是否为 Promise 类型 
           return value.then(resolve, reject) 
       } 
       setTimeout(()=>{//保证执行顺序
        if (that.state === PENDING) { //只有状态为等待中才可以进行状态改变并遍历回调函数数组
            that.state = RESOLVED 
            that.value = value 
            that.resolvedCallbacks.map(cb => cb(that.value)) //或者用forEach
            console.log('resolve',value)
        }
       },0)
     } 
     //reject函数
     function reject(value) { 
         setTimeout(()=>{
            if (that.state === PENDING) {
                that.state = REJECTED 
                that.value = value 
                that.rejectedCallbacks.map(cb => cb(that.value))
            }
         },0)
    } 
```
+ then需要支持链式调用，==then返回的值需要被包裹为promise形式==，才能继续调用then方法
+ promise2 = promise1.then(onFulfilled, onRejected); onFulfilled 或 onRejected 执行的结果为x,函数时会返回一个 x，并且执行 Promise 解决过程调用 resolvePromise，这是为了不同的 Promise 都可以兼容使 用，比如 JQuery 的 Promise 能兼容 ES6 的 Promise
+ 如果 onFulfilled 或者 onRejected 执行时抛出异常e,promise2需要被reject
+ 如果 onFulfilled 不是一个函数，promise2 以promise1的值fulfilled
+ 如果 onRejected 不是一个函数，promise2 以promise1的reason rejected
+ 如果不用setTimeout这种方式的话，若Promise里面的代码是同步代码，在执行到reject或者resolve的时候，还没有执行then，所以数组里还没有值，这个时候调用的话不会报错但是不会输出任何结果，用setTimeout转为异步的话，会先去执行then方法，将回调收集到数组里，然后再去执行异步任务，这个时候就有值了。

```
Promise的执行顺序是new Promise -> then()收集回调 -> resolve/reject执行回调，这一顺序是建立在executor是异步任务的前提上的，如果executor是一个同步任务，那么顺序就会变成new Promise -> resolve/reject执行回调 -> then()收集回调，resolve的执行跑到then之前去了，为了兼容这种情况，我们给resolve/reject执行回调的操作包一个setTimeout，让它异步执行。
```


```
if (that.state === PENDING) { 
         // that.resolvedCallbacks.push(onFulfilled) 
         return ( promise2 = new MyPromise((resolve, reject) => { 
             that.resolvedCallbacks.push(() => { 
                 try { 
                     const x = onFulfilled(that.value) 
                     resolutionProcedure(promise2, x, resolve, reject)
                    } catch (r) { 
                        reject(r) 
                    } 
                }) 
            // that.rejectedCallbacks.push(onRejected) 
            that.rejectedCallbacks.push(() => { 
                try { 
                    const x = onRejected(that.value)
                    resolutionProcedure(promise2, x, resolve, reject)
                } catch (r) { 
                    reject(r) 
                } 
         }) 
         }))
         console.log('then pending')
    } 
```
对resolve和reject的操作类似，以下几个点
+ 返回promise形式的值
+ 用setTimeout和try...catch
+ x为执行onFulfilled或者onRejected的值，再调用resolutionProcedure(promise2, x, resolve, reject)

```
if (that.state === RESOLVED) { //如果状态为resolved，执行函数
        return (promise2 = new MyPromise((resolve, reject) => { 
            setTimeout(() => { 
                 try { 
                     const x = onFulfilled(that.value) 
                     resolutionProcedure(promise2, x, resolve, reject) 
                    } catch (reason) { 
                        reject(reason) 
                    } 
                 })
            })) 
        }

    // if (that.state === REJECTED) {
    //     onRejected(that.value) 
    // } 
    if (that.state === REJECTED) { //如果状态为resolved，执行函数
        return (promise2 = new MyPromise((resolve, reject) => { 
            setTimeout(() => { 
                 try { 
                     const x = onRejected(that.value) 
                     resolutionProcedure(promise2, x, resolve, reject) 
                    } catch (reason) { 
                        reject(reason) 
                    } 
                 })
            })) 
    }
```
+ then得到的值需要进一步统一处理，实现兼容多种 Promise 的 resolutionProcedure 函数

```
// promise2,then的我们需要的返回，x为then执行onFulfilled或onRejected返回值，可能是普通值，也可能是promise
//resolve,reject为promise2的对应resolve和reject
function resolutionProcedure(promise2, x, resolve, reject) { 
    if (promise2 === x) { 
        return reject(new TypeError('Error')) 
    } 
}
//避免巡回引用问题，promise2与x不能相等
let p = new MyPromise((resolve, reject) => { 
resolve(1)
})
let p1 = p.then(value => { return p1})
```
+ 进一步判断x的类型
+ 如果x为普通值，直接resolve
+ 如果x为promise，那么它必然有一个then方法，判断x.then的类型，如果是普通值（then链结束），resolve处理，如果为function，那么调用它的then方法，并改变调用次数的值，然后递归调用，知道最终成为一个普通值

```
// promise2,then的我们需要的返回，x为then执行onFulfilled或onRejected返回值，可能是普通值，也可能是promise
//resolve,reject为promise2的对应resolve和reject
function resolutionProcedure(promise2, x, resolve, reject) { 
    if (promise2 === x) { 
        return reject(new TypeError('Error')) 
    } 
    // if (x instanceof MyPromise) { 
    //      x.then(function(value) { 
    //         resolutionProcedure(promise2, value, resolve, reject) 
    //     }, reject) 
    // } 
    let called = false //首先创建一个变量 called 用于判断是否已经调用过函数
    if (x !== null && (typeof x === 'object' || typeof x === 'function')) { //判断为是函数或者对象（promise）
         try { 
            // 判断得到x为函数或者对象（promise类型）
            let then = x.then //判断类型
            if (typeof then === 'function') { //判断得到x.then为函数，说明它是promise里的then方法
                then.call(  x, y => { //使用x.then方法
                                   if (called) return //如果已经调用过，直接返回
                                    called = true //没有调用过，标记为调用过
                                    resolutionProcedure(promise2, y, resolve, reject) //由于成功的返回y还可能是promise，所以需要递归
                                }, e => {  
                                    if (called) return 
                                    called = true 
                                    reject(e) 
                                 }) 
             } else { //判断得到x.then不为函数，直接resolve
                resolve(x)
             } 
         } catch (e) { 
             if (called) return 
             called = true 
             reject(e) 
             } 
    } else { //不是对象或者函数（不是promise，值直接传入resolve中）
        resolve(x) 
    } 
}
```
##### 完整版Promise A+

```
        const PENDING='pending'
        const RESOLVED='resolved'
        const REJECTED='rejected'
        function MyPromise(fn){
            const that = this
            that.state= PENDING
            that.resolvedCallbacks=[]
            that.rejectedCallbacks=[]
            that.value=null
            function resolve(value) { 
                if (value instanceof MyPromise) {//首先需要判断传入的值是否为 Promise 类型 
                    return value.then(resolve, reject) 
                } 
                setTimeout(()=>{//保证执行顺序
                    if (that.state === PENDING) { //只有状态为等待中才可以进行状态改变并遍历回调函数数组
                        that.state = RESOLVED 
                        that.value = value 
                        that.resolvedCallbacks.map(cb => cb(that.value)) //或者用forEach
                    }
                },0)
                } 
                //reject函数
                function reject(value) { 
                    setTimeout(()=>{
                        if (that.state === PENDING) {
                            that.state = REJECTED 
                            that.value = value 
                            that.rejectedCallbacks.map(cb => cb(that.value))
                        }
                    },0)
                } 
            try{
                fn(resolve,reject)
            }catch(e){
                reject(e)
            }
        }

        MyPromise.prototype.then=function(onFulfilled,onRejected){
            const that =this
            onFulfilled = typeof onFulfilled === 'function' ? onFulfilled : v => v 
            onRejected = typeof onRejected === 'function' ? onRejected : r => {throw r } 
            if (that.state === PENDING) { 
                // that.resolvedCallbacks.push(onFulfilled) 
                return ( promise2 = new MyPromise((resolve, reject) => { 
                    that.resolvedCallbacks.push(() => { 
                        try { 
                            const x = onFulfilled(that.value) 
                            resolutionProcedure(promise2, x, resolve, reject)
                            } catch (r) { 
                                reject(r) 
                            } 
                        }) 
                    // that.rejectedCallbacks.push(onRejected) 
                    that.rejectedCallbacks.push(() => { 
                        try { 
                            const x = onRejected(that.value)
                            resolutionProcedure(promise2, x, resolve, reject)
                        } catch (r) { 
                            reject(r) 
                        } 
                }) 
                }))
            } 
            if (that.state === RESOLVED) { //如果状态为resolved，执行函数
                return (promise2 = new MyPromise((resolve, reject) => { 
                    setTimeout(() => { 
                        try { 
                            const x = onFulfilled(that.value) 
                            resolutionProcedure(promise2, x, resolve, reject) 
                            } catch (reason) { 
                                reject(reason) 
                            } 
                        })
                    })) 
                }

            // if (that.state === REJECTED) {
            //     onRejected(that.value) 
            // } 
            if (that.state === REJECTED) { //如果状态为resolved，执行函数
                return (promise2 = new MyPromise((resolve, reject) => { 
                    setTimeout(() => { 
                        try { 
                            const x = onRejected(that.value) 
                            resolutionProcedure(promise2, x, resolve, reject) 
                            } catch (reason) { 
                                reject(reason) 
                            } 
                        })
                    })) 
            }

        } 
// promise2,then的我们需要的返回，x为then执行onFulfilled或onRejected返回值，可能是普通值，也可能是promise
//resolve,reject为promise2的对应resolve和reject
function resolutionProcedure(promise2, x, resolve, reject) { 
    if (promise2 === x) { 
        return reject(new TypeError('Error')) 
    } 
    // if (x instanceof MyPromise) { 
    //      x.then(function(value) { 
    //         resolutionProcedure(promise2, value, resolve, reject) 
    //     }, reject) 
    // } 
    let called = false //首先创建一个变量 called 用于判断是否已经调用过函数
    if (x !== null && (typeof x === 'object' || typeof x === 'function')) { //判断为是函数或者对象（promise）
         try { 
            // 判断得到x为函数或者对象（promise类型）
            let then = x.then //判断类型
            if (typeof then === 'function') { //判断得到x.then为函数，说明它是promise里的then方法
                then.call(  x, y => { //使用x.then方法
                                   if (called) return //如果已经调用过，直接返回
                                    called = true //没有调用过，标记为调用过
                                    resolutionProcedure(promise2, y, resolve, reject) //由于成功的返回y还可能是promise，所以需要递归
                                }, e => {  
                                    if (called) return 
                                    called = true 
                                    reject(e) 
                                 }) 
             } else { //判断得到x.then不为函数，直接resolve
                resolve(x)
             } 
         } catch (e) { 
             if (called) return 
             called = true 
             reject(e) 
             } 
    } else { //不是对象或者函数（不是promise，值直接传入resolve中）
        resolve(x) 
    } 
}

const p1 = new MyPromise((resolve, reject) => {
  resolve(1)          //同步executor测试
}).then(res => {
    console.log(res)
    return 2          //链式调用测试
  })

  .then()             //值穿透测试
  .then(res => {
    console.log(res)
    return new MyPromise((resolve, reject) => {
      resolve(3)      //返回Promise测试
    })
  })
  .then(res => {
    console.log(res)
    throw new Error('reject测试')   //reject测试
  })
  .then(() => {}, err => {
    console.log(err)
  })
// 1 
// 2 
// 3 
// Error: reject测试
```
##### 实现 promiseify 方法
作用：把异步函数转换为promise,直接可以使用then

```
let Promise = require('./bluebird');
let fs = require("fs");

var readFile = Promise.promisify(fs.readFile);
readFile("1.txt", "utf8").then(function(data) {
    console.log(data);
})
```
写法

```
MyPromise.promisify = function(fn) {
    return function() {
        var args = Array.from(arguments);
        return new MyPromise(function(resolve, reject) {
            fn.apply(null, args.concat(function(err) {
                err ? reject(err) : resolve(arguments[1])
            }));
        })
    }
}
```
+ 返回一个promise
+ 内部将原来参数后面拼接一个回调函数参数，在回调函数里执行这个promise的reslove方法把结果传出去，promiseify就实现了

##### 手写race方法
+ **Promise.resolve(p)**用于处理传入值不为Promise的情况
+ resolve是上边new的promise
+ 直接forEach即可
```
let race=function(arr){
  return new Promise((resolve,reject)=>{
    arr.forEach(item=>{
      Promise.resolve(item).then(resolve,reject)
    })
  })
}

let p1 = new Promise((resolve, reject) => {
    setTimeout(() => {
        resolve('success2')
    }, 1000)
})
let p2 = new Promise((resolve, reject) => {
    setTimeout(() => {
        reject('reject')
    }, 500)
})
// 调用
race([p1, p2]).then(res => {
    console.log(res);
}).catch(err => {
    console.error(err);
})
```

```
//race方法 
static race(promiseArr) {
  return new MyPromise((resolve, reject) => {
    //同时执行Promise,如果有一个Promise的状态发生改变,就变更新MyPromise的状态
    for (let p of promiseArr) {
      MyPromise.resolve(p).then(  //Promise.resolve(p)用于处理传入值不为Promise的情况
        value => {
          resolve(value)        //注意这个resolve是上边new MyPromise的
        },
        err => {
          reject(err)
        }
      )
    }
  })
}
```
##### 手写all方法
+ 当所有的任务都执行完成时，Promise.all()返回resolve，但当有一个失败(reject)，则返回失败的信息，即使其他promise执行成功，也会返回失败。
+ 需要先把所有方法搜集到一个数组里
+ promise.resolve()包裹
```
//静态的all方法
static all(promiseArr) {
  let index = 0
  let result = []
  return new MyPromise((resolve, reject) => {
    promiseArr.forEach((p, i) => {
      //Promise.resolve(p)用于处理传入值不为Promise的情况
      MyPromise.resolve(p).then(
        val => {
          index++
          result[i] = val
          //所有then执行后, resolve结果
          if(index === promiseArr.length) {
            resolve(result)
          }
        },
        err => {
          //有一个Promise被reject时，MyPromise的状态变为reject
          reject(err)
        }
      )
    })
  })
}
```
##### 手写finally
finally()方法返回一个Promise。在promise结束时，无论结果是fulfilled或者是rejected，都会执行指定的回调函数。在finally之后，还可以继续then。并且会将值原封不动的传递给后面的then

```
Promise.prototype.finally = function (callback) {
  let P = this.constructor;
  return this.then(
    value  => P.resolve(callback()).then(() => value),
    reason => P.resolve(callback()).then(() => { throw reason })
  );
};
```

```
为什么需要Promise.resolve(callback()).then(() => value)
而不能直接执行callback, return value

因为callback如果是个异步操作，返回promise呢.希望等callback执行完再接着执行
```



##### catch,all,resolve,reject,race

```
//catch方法其实就是执行一下then的第二个回调
catch(rejectFn) {
  return this.then(undefined, rejectFn)
}
```

```
//静态的resolve方法
static resolve(value) {
  if(value instanceof MyPromise) return value // 根据规范, 如果参数是Promise实例, 直接return这个实例
  return new MyPromise(resolve => resolve(value))
}
```

```
//静态的reject方法
static reject(reason) {
  return new MyPromise((resolve, reject) => reject(reason))
}
```






### 5.Generator
生成器，可以解决回调地狱的问题，可以在函数的某些部分停止

```
function *foo(x) { 
  let y = 2 * (yield (x + 1)) 
  let z = yield (y / 3) 
  return (x + y + z) 
}
let it = foo(5)
console.log(it.next())   // => {value: 6, done: false}
console.log(it.next(12)) // => {value: 8, done: false}
console.log(it.next(13)) // => {value: 42, done: true}
```
+ 第一次调用next,停在(yield (x + 1))，输出为6
+ 第二次调用next，yield输入参数12，则y=2*12，停在yield (y / 3) ，也就是24/3=8
+ 第三次调用next,yield传入13，则z=13，y=24,x=5,和为42

#### 解决回调地狱
封装ajax
```
function *fetch() {
    yield ajax(url, () => {})
    yield ajax(url1, () => {})
    yield ajax(url2, () => {})
}
let it = fetch()
let result1 = it.next()
let result2 = it.next()
let result3 = it.next()
```
### 6.解决Generator的获取控制权问题
Generator函数的缺点，必须通过yield停止程序，再通过next()把新的值传入才能继续执行，需要有一种前一步执行结束把数据和控制权交给下一步的机制（Thunk,和promise的then）

```
var fs = require('fs');
var thunkify = require('thunkify');
var readFile = thunkify(fs.readFile);

var gen = function* (){
  var r1 = yield readFile('/etc/fstab');
  console.log(r1.toString());
  var r2 = yield readFile('/etc/shells');
  console.log(r2.toString());
};
//取得函数的控制权必须通过next(),而且需要把上一步的值再次传入
var g = gen();
var r1 = g.next();
r1.value(function(err, data){
  if (err) throw err;
  var r2 = g.next(data);
  r2.value(function(err, data){
    if (err) throw err;
    g.next(data);
  });
});
```
#### Thunk函数

```
传值调用，也就是表达式在传入之前先求值，再调用

f(x + 5)
// 传值调用时，等同于
f(6)
```

```
传名调用，先传入所有值，调用的时候再计算
f(x + 5)
// 传名调用时，等同于
(x + 5) * 2
```
JS是传值调用的，转换为传名调用，也就是Thunk函数(有些类似<a href="###函数柯里化">函数柯里化</a>)

```
var Thunk = function(fn){
  return function (){
    var args = Array.prototype.slice.call(arguments);
    return function (callback){
      args.push(callback);
      return fn.apply(this, args);
    }
  };
};
```
效果
```
// 正常版本的readFile（多参数版本）
fs.readFile(fileName, callback);
//Thunk函数处理后的效果
var readFileThunk = Thunk(fs.readFile);
readFileThunk(fileA)(callback);
```
#### Thunk函数与Generator结合控制执行

+ 自动执行，目的是让gen里读取文件的操作自动s顺序全部执行，yield之后都是Thunk函数

```
var gen = function* (){
  var f1 = yield readFile('fileA');
  var f2 = yield readFile('fileB');
  // ...
  var fn = yield readFile('fileN');
};
```
```
// 获取传入的函数，然后自动执行next(),获取next(data)的值
//如果执行完毕，直接退出
//如果没有有执行完毕，将next函数再传入Thunk函数当中（result.value），相当是将generator当前的执行中断的指针与上一步的数据 自动传入下一步
function run(fn) {
  var newfn = fn();
  function next(err, data) {
    var result = newfn.next(data);
    if (result.done) return;
    result.value(next);
  }
  next();
}
run(gen);
```
#### Promise对象结合generator控制执行

```
var fs = require('fs');
var readFile = function (fileName){
  return new Promise(function (resolve, reject){
    fs.readFile(fileName, function(error, data){
      if (error) reject(error);
      resolve(data);
    });
  });
};

var gen = function* (){
  var f1 = yield readFile('/etc/fstab');
  var f2 = yield readFile('/etc/shells');
  console.log(f1.toString());
  console.log(f2.toString());
};
```
使用，也是类似于利用then获取之前的值再传入

```
//{ value: Promise { <pending> }, done: false }，value本身为promise对象
var g = gen();
g.next().value.then(function(data){
  g.next(data).value.then(function(data){
    g.next(data);
  });
})
```
自动执行

```
function run(gen){
  var g = gen();
  function next(data){
    var result = g.next(data);
    if (result.done) return result.value;
    result.value.then(function(data){
      next(data);
    });
  }
  next();
}
run(gen);
```
#### co函数
co 函数库其实就是将两种自动执行器（Thunk 函数和 Promise 对象），包装成一个库
+ Generator 函数的 yield 命令后面，只能是 Thunk 函数或 Promise 对象

```
//使用方法
var gen = function* (){
  var f1 = yield readFile('/etc/fstab');
  var f2 = yield readFile('/etc/shells');
};
...
var co = require('co');
co(gen).then(function (){
  console.log('Generator 函数执行完成');
})
```
实现源码
+ 输入为一个generator函数
+ 输出为promise对象，如果输入的generator是函数，那么就执行，如果不是，就直接resolve包裹

```
function co(gen) {
  var ctx = this;

  return new Promise(function(resolve, reject) {
    if (typeof gen === 'function') gen = gen.call(ctx);
    if (!gen || typeof gen.next !== 'function') return resolve(gen);
  });
}
```
+ onFulfilled()方法包裹next()方法，主要用于捕获错误

```
onFulfilled();
    function onFulfilled(res) {
      var ret;
      try {
        ret = gen.next(res);
      } catch (e) {
        return reject(e);
      }
      next(ret);
    }    
```
+ 第一行，检查当前是否为 Generator 函数的最后一步，如果是就返回。
+ 第二行，确保每一步的返回值，是 Promise 对象。
+ 第三行，使用 then 方法，为返回值加上回调函数，然后通过 onFulfilled 函数再次调用 next 函数。
+ 第四行，在参数不符合要求的情况下（参数非 Thunk 函数和 Promise 对象），将 Promise 对象的状态改为 rejected，从而终止执行
```
function next(ret) {
  if (ret.done) {
      return resolve(ret.value);
  }
  var value = toPromise.call(ctx, ret.value);
  if (value && isPromise(value)){
     return value.then(onFulfilled, onRejected); 
  }
  return onRejected(new TypeError('You may only yield a function, promise, generator, array, or object, '
        + 'but the following object was passed: "' + String(ret.value) + '"'));
    }
});
```
******************************
### 7.async函数(spawn的实现)
async 函数就是将 Generator 函数的星号（*）替换成 async，将 yield 替换成 await。async 函数的实现，就是将 Generator 函数和自动执行器，包装在一个函数里。

```

async function fn(args){
 // ...
}
// 等同于
function fn(args){ 
  return spawn(function*() {
    // ...
  }); 
}
```
==spawn函数的实现==

```
function spawn(genF) {
  return new Promise(function(resolve, reject) {
    var gen = genF();
    function step(nextF) {
      try {
        var next = nextF();
      } catch(e) {
        return reject(e); 
      }
      if(next.done) {
        return resolve(next.value);
      } 
      Promise.resolve(next.value).then(function(v) {
        step(function() { return gen.next(v); });      
      }, function(e) {
        step(function() { return gen.throw(e); });
      });
    }
    step(function() { return gen.next(undefined); });
  });
}
```
********************
### 8.常见的定时器函数
setsetTimeout()和setInterval()的事件可能都不会特别准确没具体原因与JS运行机制有关
<a href="### JS运行机制">JS运行机制</a>
#### setTimeout()

```
setTimeout(function(){...},delay)
```
#### setInterval()

```
setInterval(function(){...},delay)
```
#### setTimeout()与setInterval的实现
setTimeout()方法在等待指定的毫秒数之后执行一个函数。setInterval()可以重复执行

**setTimeout()实现setInterval**

```
function mysetinterval(fn,delay){
    function interval(){
        fn();
        setTimeout(interval,delay)
    }
    setTimeout(interval,delay)
}
function myfn(){
    console.log('1')
}

mysetinterval(myfn,1000)
```


#### requestAnimationFrame

+ 浏览器每秒最多只能重绘60次，requestAnimationFrame的基本思想就是与这个刷新频率保持同步，利用这个刷新频率进行页面重绘。

```
//回调函数会在重绘之前调用
requestID = window.requestAnimationFrame(callback); 
// 取消
cancelAnimationFrame(...);
```
比如，长列表分批次进行渲染

```
var total = 100000;
var size = 100;
var count = total / size;
var done = 0;
var ul = document.getElementById('list');

function addItems() {
    var li = null;
    var fg = document.createDocumentFragment();
    for (var i = 0; i < size; i++) {
        li = document.createElement('li');
        li.innerText = 'item ' + (done * size + i);
        fg.appendChild(li);
    }
    ul.appendChild(fg);
    done++;
    if (done < count) {
        requestAnimationFrame(addItems);
    }
};
requestAnimationFrame(addItems);
```
对setInterval进行改进

```
function setInterval(callback, interval) {
  let timer
  const now = Date.now
  let startTime = now()
  let endTime = startTime
  const loop = () => {
    //重绘前调用一次loop
    timer = window.requestAnimationFrame(loop)
    //执行loop
    endTime = now()
    if (endTime - startTime >= interval) {
      startTime = endTime = now()
      callback(timer)
    }
  }
  //结束后调用下一次loop
  timer = window.requestAnimationFrame(loop)
  return timer
}
let a = 0
setInterval(timer => {
  console.log(1)
  a++
  //取消调用
  if (a === 3) cancelAnimationFrame(timer)
}, 1000)
<div class="md-section-divider"></div>
```
***********************
### 9.setTimeout和promise的区别
Promise 是微任务，setTimeout 是宏任务，同一个事件循环中，promise.then总是先于 setTimeout 执行。

# async、await 细节

资料：https://juejin.cn/post/7194744938276323384#heading-1

## async的返回值

`async`函数在抛出返回值时，会根据返回值**类型**开启**不同数目的微任务**

- return结果值：非`thenable`、非`promise`（不等待）
- return结果值：`thenable`（等待 1个`then`的时间）
- return结果值：`promise`（等待 2个`then`的时间）

```js
async function testC () {
    return new Promise((resolve, reject) => {
        resolve()
    })
} 

testC().then(() => console.log(1));
Promise.resolve()
    .then(() => console.log(2))
    .then(() => console.log(3))
    .then(() => console.log(4))
// (等待两个then)最终结果: 2 3 1 4
```

## **await 右值类型区别**

+ `await`后面接非 `thenable` 类型，会立即向微任务队列添加一个微任务`then`，**但不需等待**
+ `await` 后面接 `thenable` 类型，需要**等待一个 `then` 的时间之后**执行
+ `await` 后面接 promise 类型， 需要**等待一个 `then` 的时间之后**执行（TC 39(ECMAScript标准制定者) 对`await` 后面是 `promise` 的情况如何处理进行了一次修改，**移除**了额外的两个微任务，在**早期版本**，依然会等待两个 `then` 的时间）
+  **完全可以将 `await` 后面的代码可以看做在 `then` 里面执行的结果**

```js
function func () {
    console.log(2);
}
async function test () {
    console.log(1);
    await func();
    console.log(3);
}
test();
console.log(4);

// 最终结果👉: 1 2 4 3
```

## 其他

`await`一定要等到右侧的表达式有**确切的值**才会放行，否则将一直等待（阻塞当前`async`函数内的后续代码）

```js
function func () {
    return new Promise((resolve) => {
        console.log('B')
        // resolve() 故意一直保持pending
    })
}
async function test () {
    console.log(1);
    await func()
    console.log(3);
}
test();
console.log(4);
// 最终结果👉: 1 B 4 (永远不会打印3)
```

